ctapipe/reco/sklearn.py

```diff
@@ -0,0 +1,590 @@
+                    "Must provide `model_cls` if not loading model from file"
+                )
+
+            if subarray is None:
+                raise TypeError(
+                    "__init__() missing 1 required positional argument: 'subarray'"
+                )
+
+            super().__init__(subarray, **kwargs)
+            self.subarray = subarray
+            self.qualityquery = QualityQuery(parent=self)
+            self.generate_features = FeatureGenerator(parent=self)
+
+            # to verify settings
+            self._new_model()
+
+            self._models = {} if models is None else models
+            self.unit = None
+            self.stereo_combiner = StereoCombiner.from_name(
+                self.stereo_combiner_cls,
+                prefix=self.model_cls,
+                property=self.property,
+                parent=self,
+            )
+        else:
+            loaded = self.read(self.load_path)
+            if (
+                subarray is not None
+                and loaded.subarray.telescope_types != subarray.telescope_types
+            ):
+                self.log.warning(
+                    "Supplied subarray has different telescopes than subarray loaded from file"
+                )
+            self.__dict__.update(loaded.__dict__)
+            self.subarray = subarray
+
+    @abstractmethod
+    def __call__(self, event: ArrayEventContainer) -> None:
+        """Event-wise prediction for the EventSource-Loop.
+
+        Fills the event.dl2.<your-feature>[name] container.
+
+        Parameters
+        ----------
+        event: ArrayEventContainer
+        """
+
+    @abstractmethod
+    def predict_table(self, key, table: Table) -> Table:
+        """
+        Predict on a table of events
+
+        Parameters
+        ----------
+        table : `~astropy.table.Table`
+            Table of features
+
+        Returns
+        -------
+        table : `~astropy.table.Table`
+            Table(s) with predictions, matches the corresponding
+            container definition(s)
+        """
+
+    def write(self, path):
+        Provenance().add_output_file(path, role="ml-models")
+        with open(path, "wb") as f:
+            joblib.dump(self, f, compress=True)
+
+    @classmethod
+    def read(cls, path, **kwargs):
+        with open(path, "rb") as f:
+            instance = joblib.load(f)
+
+        for attr, value in kwargs.items():
+            setattr(instance, attr, value)
+
+        if not isinstance(instance, cls):
+            raise TypeError(
+                f"{path} did not contain an instance of {cls}, got {instance}"
+            )
+
+        Provenance().add_input_file(path, role="ml-model")
+        return instance
+
+    @lazyproperty
+    def instrument_table(self):
+        return QTable(self.subarray.to_table("joined"))
+
+    def _new_model(self):
+        return SUPPORTED_MODELS[self.model_cls](**self.model_config)
+
+    def _table_to_X(self, table):
```

Comment:
Although these `table_to` are only used in the fit method below, we might want to document them. It might not be completely obvious why they behave so differently.
Also `valid` might indicate it is the subset of valid `X` as opposed to a boolean mask. Later on we also call it `mask`

I understand the code, but I have also worked on/with it for a while, not sure how clear this is for other people in the future.